### PR TITLE
[ASR Pass] Implement `sqrt` and `dsqrt` in IntrinsicScalarFunction

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Fixes: https://github.com/lfortran/lfortran/issues/2770

### CODE REVIEW

**Comment by certik** [comment]
The failure is https://github.com/lfortran/lfortran/issues/2776.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -1456,6 +1456,9 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
             return;
         }
         this->visit_expr(*x.m_arg);
+        if (tmp->getType()->isPointerTy()) {
+            tmp = CreateLoad(tmp);
+        }
```
Comment (certik) [inline src/libasr/codegen/asr_to_llvm.cpp:1461]: Let's remove `visit_IntrinsicFunctionSqrt`? We don't use it anymore.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 2]
```diff
@@ -1658,6 +1662,109 @@ namespace Aint {
...
+            args.p, args.n, 0, return_type, m_value);
+    }
+
+    static inline ASR::expr_t* instantiate_Sqrt(Allocator &al, const Location &loc,
+            SymbolTable *scope, Vec<ASR::ttype_t*>& arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t>& new_args, int64_t overload_id) {
+        ASR::ttype_t* arg_type = arg_types[0];
+        if (is_real(*arg_type)) {
+            return EXPR(ASR::make_IntrinsicFunctionSqrt_t(al, loc,
+                new_args[0].m_value, return_type, nullptr));
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:1719]: Why is this needed? I think we might need a dedicated node, but it should be called `RealSqrt`.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 1 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_llvm.cpp:1461]: If we remove `IntrinsicFunctionSqrt`, we have add a new C interface for real sqrt functions. 
Here, as it uses `llvm.sqrt.f32`, I thought let it be.
Should I remove it completely?

File: src/libasr/pass/intrinsic_function_registry.h [hunk 3]
```diff
@@ -1658,6 +1662,109 @@ namespace Aint {
...
+        } else {
+            return UnaryIntrinsicFunction::instantiate_functions(al, loc, scope,
+                "sqrt", arg_type, return_type, new_args, overload_id);
+        }
+    }
+
+}  // namespace Sqrt
+
+namespace Dsqrt {
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:1728]: Why do we need `Dsqrt`?

We should have only one function `sqrt`, that can accept single and double precision arguments.

File: src/libasr/codegen/asr_to_llvm.cpp (hunk 1 above)
Comment (certik) [reply src/libasr/codegen/asr_to_llvm.cpp:1461]: Let's leave it then. I would (later) rename `IntrinsicFunctionSqrt` to `RealSqrt`, since it's just a special operation on reals, implemented directly in the backend.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 2 above)
Comment (certik) [reply src/libasr/pass/intrinsic_function_registry.h:1719]: So this is fine now.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 3 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:1728]: I think dsqrt accepts only double precision, so I added it separately.
https://gcc.gnu.org/onlinedocs/gcc-3.4.6/g77/DSqRt-Intrinsic.html
Here, `kind=2` means 8, I think. I tested it in GFortran, it only works for double precision.

Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:1728]: Should we redesign IntrinsicScalarFunction to implement both of them in one function?
Apart from redesigning, I'm not able to see how we can proceed.

Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:1728]: I can do the following changes, but we can't throw error for single precision passed as argument to `dsqrt`
```diff
diff --git a/src/libasr/pass/intrinsic_function_registry.h b/src/libasr/pass/intrinsic_function_registry.h
index a05a6bda7..0e8001391 100644
--- a/src/libasr/pass/intrinsic_function_registry.h
+++ b/src/libasr/pass/intrinsic_function_registry.h
@@ -3527,7 +3527,7 @@ namespace IntrinsicScalarFunctionRegistry {
                 {"radix", {&Radix::create_Radix, nullptr}},
                 {"sign", {&Sign::create_Sign, &Sign::eval_Sign}},
                 {"aint", {&Aint::create_Aint, &Aint::eval_Aint}},
-                {"dsqrt", {&Dsqrt::create_Dsqrt, &Dsqrt::eval_Dsqrt}},
+                {"dsqrt", {&Sqrt::create_Sqrt, &Sqrt::eval_Sqrt}},
                 {"sqrt", {&Sqrt::create_Sqrt, &Sqrt::eval_Sqrt}},
                 {"sngl", {&Sngl::create_Sngl, &Sngl::eval_Sngl}},
                 {"Symbol", {&SymbolicSymbol::create_SymbolicSymbol, &SymbolicSymbol::eval_SymbolicSymbol}},
```

**Comment by certik** [comment]
Regarding the `sqrt` vs `dsqrt`:

So Fortran historically had `sin` (single precision) and `dsin` (double precision). Then in 1990 it introduced generic function `sin` that can handle any precision. For backwards compatibility it kept `dsin`. So you can do `sin(1.0_dp)` and `dsin(1.0_dp)` and it is equivalent.

In ASR, we only have `sin`, implemented as `IntrinsicScalarFunction`. We have `overload_id` in there, which we can use (I am not sure if we do currently) to distinguish the argument type, such as single, double quadruple precision, and complex single/double (and even symbolic "sin" later on), etc. This can be useful in the backend or ASR intrinsic function pass to quickly generate the proper `sin` implementation.

`sqrt` should behave just like `sin`.

The only caveat is that most backends have a dedicated instruction for sqrt (I believe both LLVM and WASM) that we want to use, for performance. Since `RealSqrt` is such a fundamental operation, we might just add it as a dedicated ASR node. In general however, say some future CPU will have a dedicated instruction to compute `sin` (the Intel CPU has it, but its slower than our hand implementation). The design of `IntrinsicScalarFunction` is such that in the ASR intrinsic pass the LLVM backend could select which functions it wants the pass to instantiate (implement), and which functions to leave. So we could represent `Sqrt` using `IntrinsicScalarFunction`, and the LLVM backend would select to skip `Sqrt`, as it would implement it iself. We will for sure use this approach for other functions.

So either of these two designs is fine.

Regarding `dsqrt`, that should just be represented as "sqrt" in ASR, whether `RealSqrt` or `IntrinsicScalarFunction`.

**Comment by Thirumalai-Shaktivel** [comment]
Ready for review!

File: src/libasr/pass/intrinsic_function_registry.h [hunk 4]
```diff
@@ -61,6 +61,8 @@ enum class IntrinsicScalarFunctions : int64_t {
     Sign,
     SignFromValue,
     Aint,
+    Sqrt,
+    DSqrt,
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:65]: DSqrt should not be a function here.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 5]
```diff
@@ -3412,6 +3487,8 @@ namespace IntrinsicScalarFunctionRegistry {
                 {"radix", {&Radix::create_Radix, nullptr}},
                 {"sign", {&Sign::create_Sign, &Sign::eval_Sign}},
                 {"aint", {&Aint::create_Aint, &Aint::eval_Aint}},
+                {"sqrt", {&Sqrt::create_Sqrt, &Sqrt::eval_Sqrt}},
+                {"dsqrt", {&Sqrt::create_Sqrt, &Sqrt::eval_Sqrt}},
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:3491]: How is `dsin` handled? It's not mentioned here, so I would not mention `dsqrt` either, although I am not sure how this mapping is handled.

**Comment by certik** [comment]
I think this looks quite good now. I left some comments.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 5 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:3491]: Okay, I got it now!
double precision intrinsics are mapped in the `ast_common_visitor`. I didn't know about this.
I will make the required changes, thanks for the review

**Review by certik — APPROVED** [review]
I think it looks good now. Thanks!

**Comment by Thirumalai-Shaktivel** [comment]
Thanks for the review!
